src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAProjectWithAllFeatures.cs

```diff
@@ -99,14 +81,94 @@ public void It_publishes_the_project_correctly(string targetFramework, string []
     }
 }");
             baselineConfigJsonObject["runtimeOptions"]["tfm"] = targetFramework;
-            baselineConfigJsonObject["runtimeOptions"]["framework"]["version"] = 
+            baselineConfigJsonObject["runtimeOptions"]["framework"]["version"] =
                 targetFramework == "netcoreapp1.0" ? "1.0.5" : "1.1.2";
 
             runtimeConfigJsonObject
                 .Should()
                 .BeEquivalentTo(baselineConfigJsonObject);
         }
 
+        [Fact]
+        public void It_fails_when_nobuild_is_set_and_build_was_not_performed_previously()
+        {
+            var publishCommand = GetPublishCommand("netcoreapp1.0").Execute("/p:NoBuild=true");
+            publishCommand.Should().Fail().And.HaveStdOutContaining("MSB3030"); // "Could not copy ___ because it was not found."
+        }
+
+        [Theory]
+        [MemberData(nameof(PublishData))]
+        public void It_does_not_build_when_nobuild_is_set(string targetFramework, string[] expectedPublishFiles)
+        {
+            var publishCommand = GetPublishCommand(targetFramework);
+
+            // do a separate build invocation before publish
+            var buildCommand = new BuildCommand(Log, publishCommand.ProjectRootPath);
+            buildCommand.Execute().Should().Pass();
+
+            // modify all project files, which would force recompilation if we were to build during publish
+            foreach (string projectFile in EnumerateFiles(buildCommand, "*.csproj"))
+            {
+                File.AppendAllText(projectFile, " ");
+            }
+
+            // capture modification time of all binaries before publish
+            var modificationTimes = GetLastWriteTimesUtc(buildCommand, "*.exe", "*.dll", "*.resources", "*.pdb");
+
+            // publish (with NoBuild set)
+            publishCommand.Execute("/p:NoBuild=true").Should().Pass();
```

Comment:
As a sanity check for this test, have you tried running it without the `/p:NoBuild=true` parameter and verified that it fails due to updated modification dates of the output files?

Reply:
Yes, I have. :) I also tried various permutations of not having the right targets in _BeforePublishTargets.